### PR TITLE
Adding `dry-run` option to `raft-ann-bench`

### PR DESCRIPTION
**Pull request by cjnolet**
The new `dry-run` mode will create a json config file from the rest of the input params and yaml configs. It will also print the command(s) to run the underlying C++ binaries.

### CODE REVIEW

**Comment by cjnolet** [comment]
cc @tfeher this also adds formatting to the generated json config so it's readable and editable. Hopefully this makes life easier when working w/ the underlying C++ binaries.

File: docs/source/raft_ann_benchmarks.md [hunk 1]
```diff
@@ -152,6 +152,11 @@ options:
   -f, --force           re-run algorithms even if their results already exist (default: False)
   -m SEARCH_MODE, --search-mode SEARCH_MODE
                         run search in 'latency' (measure individual batches) or 'throughput' (pipeline batches and measure end-to-end) mode (default: throughput)
+  -t SEARCH_THREADS, --search-threads SEARCH_THREADS
+                        specify the number threads to use for throughput benchmark. Single value or a pair of min and max separated by ':'. Example --threads=1:4. Power of 2 values between 'min' and 'max' will be used. If only 'min' is
```
Comment (divyegala) [inline docs/source/raft_ann_benchmarks.md:156]: Example should be `--search-threads`

File: python/raft-ann-bench/src/raft-ann-bench/run/__main__.py [hunk 2]
```diff
@@ -271,6 +303,11 @@ def main():
         build = args.build
         search = args.search
 
+    if not args.dry_run:
+        dry_run = False
+    else:
+        dry_run = True
```
Comment (divyegala) [inline python/raft-ann-bench/src/raft-ann-bench/run/__main__.py:309]:
```suggestion
    dry_run = args.dry_run
```

File: python/raft-ann-bench/src/raft-ann-bench/run/__main__.py [hunk 3]
```diff
@@ -80,23 +81,24 @@ def run_build_and_search(
...
     batch_size,
     search_threads,
     mode="throughput",
 ):
     for executable, ann_executable_path, algo in executables_to_run.keys():
         # Need to write temporary configuration
-        temp_conf_filename = f"temporary_{conf_filename}"
-        temp_conf_filepath = os.path.join(conf_filedir, temp_conf_filename)
-        with open(temp_conf_filepath, "w") as f:
+        temp_conf_filename = f"temporary_{conf_filename}_{uuid.uuid1()}.json"
```
Comment (divyegala) [inline python/raft-ann-bench/src/raft-ann-bench/run/__main__.py:92]: Let's change this to
```suggestion
        temp_conf_filename = f"{conf_filename}_{algo}_{uuid.uuid1()}.json"
```

File: docs/source/raft_ann_benchmarks.md (hunk 1 above)
Comment (cjnolet) [reply docs/source/raft_ann_benchmarks.md:156]: ooh good catch. This was copied from the c++ binary.

Comment (divyegala) [reply docs/source/raft_ann_benchmarks.md:156]: I guess the script's help section will also have to be updated?

Comment (cjnolet) [reply docs/source/raft_ann_benchmarks.md]: Yeah, I updated that as well

**Review by divyegala — APPROVED** [review]

**Pull request merged**